### PR TITLE
CSI Topology ga

### PR DESCRIPTION
**Pull request by msau42**
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:
Move CSI Topology feature to GA. This PR adds the CSINode object to storage.k8s.io/v1 API group.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
CSI Topology feature is GA. The CSINodeInfo feature gate is deprecated and will be removed in a future release. The storage.k8s.io/v1beta1 CSINode object is deprecated and will be removed in a future release.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/ac738ac78ec2e425c54354119526bac916baeb7e/keps/sig-storage/20190124-csi-topology.md
- [Usage]: https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode
- [Other doc]: https://kubernetes-csi.github.io/docs/topology.html
```

### CODE REVIEW

File: staging/src/k8s.io/api/storage/v1beta1/types.go [hunk 1]
```diff
@@ -348,6 +348,8 @@ const (
 // +genclient:nonNamespaced
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// DEPRECATED - This group version of CSINode is deprecated by storage/v1/CSINode.
```
Comment (msau42) [inline staging/src/k8s.io/api/storage/v1beta1/types.go]: Side note: I noticed the other objects in here, StorageClass, VolumeAttachment are not deprecated. Should we go ahead and deprecate them (in a separate PR)?

File: test/integration/etcd/data.go [hunk 2]
```diff
@@ -544,13 +544,20 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
...
 		etcdStorageData[gvr("storage.k8s.io", "v1beta1", "csinodes")] = StorageData{
 			Stub:             `{"metadata": {"name": "csini1"}, "spec": {"drivers": [{"name": "test-driver", "nodeID": "localhost", "topologyKeys": ["company.com/zone1", "company.com/zone2"]}]}}`,
 			ExpectedEtcdPath: "/registry/csinodes/csini1",
 		}
+
+		// k8s.io/kubernetes/pkg/apis/storage/v1
+		etcdStorageData[gvr("storage.k8s.io", "v1", "csinodes")] = StorageData{
+			Stub:             `{"metadata": {"name": "csini2"}, "spec": {"drivers": [{"name": "test-driver", "nodeID": "localhost", "topologyKeys": ["company.com/zone1", "company.com/zone2"]}]}}`,
+			ExpectedEtcdPath: "/registry/csinodes/csini2",
+			ExpectedGVK:      gvkP("storage.k8s.io", "v1beta1", "CSINode"),
```
Comment (msau42) [inline test/integration/etcd/data.go]: Is this right? I'm not actually sure where this object-specific ordering is specified. Since the storage/v1 group already exists, all the places I found where you specify ordering already prioritize the v1 group over v1beta1. Is [this](https://github.com/kubernetes/kubernetes/blob/14344b57e56258e87cbe80c8cd80399855eca424/pkg/kubeapiserver/default_storage_factory_builder.go#L60) relevant?

**Comment by msau42** [comment]
@kubernetes/sig-storage-pr-reviews 
/priority important-soon

**Comment by liggitt** [comment]
/assign @thockin 

for context from beta review (https://github.com/kubernetes/kubernetes/pull/74283)

File: staging/src/k8s.io/kubectl/pkg/generated/bindata.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Package generated Code generated by go-bindata. (@generated) DO NOT EDIT.
+// Code generated by go-bindata.
```
Comment (msau42) [inline staging/src/k8s.io/kubectl/pkg/generated/bindata.go]: This file showed up when I ran "make update". Unsure if I should be including it here, it seems unrelated

Comment (msau42) [reply staging/src/k8s.io/kubectl/pkg/generated/bindata.go]: apparently I should not include it

File: staging/src/k8s.io/api/storage/v1/types.go [hunk 4]
```diff
@@ -216,3 +216,97 @@ type VolumeError struct {
...
+	// +optional
+	Allocatable *VolumeNodeResources `json:"allocatable,omitempty" protobuf:"bytes,4,opt,name=allocatable"`
+}
+
+// VolumeNodeResources is a set of resource limits for scheduling of volumes.
+type VolumeNodeResources struct {
+	// Maximum number of unique volumes managed by the CSI driver that can be used on a node.
+	// A volume that is both attached and mounted on a node is considered to be used once, not twice.
+	// The same rule applies for a unique volume that is shared among multiple pods on the same node.
+	// If this field is nil, then the supported number of volumes on this node is unbounded.
```
Comment (thockin) [inline staging/src/k8s.io/api/storage/v1/types.go]: s/is nil/is not specified/

Comment (thockin) [reply staging/src/k8s.io/api/storage/v1/types.go]: same anywhere else

File: staging/src/k8s.io/api/storage/v1beta1/types.go (hunk 1 above)
Comment (thockin) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: Originally, the plan was to deprecate whole group-versions, but that seems to have shifted in machinery.

@liggitt or @lavalamp - thoughts?  Should we deprecate individual types from a group?  I don't think we can, really.

**Review by thockin — COMMENTED** [review]
LGTM overall, just a couple nits

**Comment by liggitt** [comment]
> @liggitt or @lavalamp - thoughts? Should we deprecate individual types from a group? I don't think we can, really.

Not sure why we couldn't. API groups can be sparse, and things in the same group can progress to v1 at different rates. I would actually have expected the API types that already had `storage.k8s.io/v1` versions (StorageClass, VolumeAttachment) to already have been marked deprecated.

File: staging/src/k8s.io/api/storage/v1/types.go (hunk 4 above)
Comment (msau42) [reply staging/src/k8s.io/api/storage/v1/types.go]: updated. will fix comments on existing objects in a separate pr

File: staging/src/k8s.io/api/storage/v1beta1/types.go (hunk 1 above)
Comment (msau42) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: I'll go ahead and deprecate the other objects in a separate pr

Comment (thockin) [reply staging/src/k8s.io/api/storage/v1beta1/types.go]: I'm not sure I buy this.  At all.  Deprecation policy says:

Rule #1: API elements may only be removed by incrementing the version of the API group.

**Comment by thockin** [comment]
@liggitt I don't buy the depreaction argument, at least as per today's rules:

Rule #1: API elements may only be removed by incrementing the version of the API group.

**Comment by lavalamp** [comment]
deprecation is not removal?

**Comment by liggitt** [comment]
> Rule #1: API elements may only be removed by incrementing the version of the API group.

didn't `storage.k8s.io/v1` do that?

**Comment by msau42** [comment]
Since the original storage/v1, we added new types to v1beta1 that don't exist yet. I assume only once all the v1beta1 types are promoted to v1, then we can actually remove v1beta1.  But I guess the question is can we deprecate a single type, or do we have to wait and deprecate the entire group?

**Comment by thockin** [comment]
What is the point of deprecating it from v1beta1 if it is not actually
being removed in any subsequent version?

Frankly, this "sparse API group" stuff feels bass-ackwards to me, and
I really think we need to rethink the whole API alpha/beta stuff
top-to-bottom.

**Comment by liggitt** [comment]
> What is the point of deprecating it from v1beta1 if it is not actually being removed in any subsequent version?

Some scenarios we've encountered:

Improvements that require changing API behavior must only be applied to subsequent versions:
* Improved defaults (like deployment's spec.revisionHistoryLimit)
* Structural improvements (like ingress v1)

Bug fixes that change API behavior can generally only be fixed in subsequent versions:
* Improved validation (like better validation around custom resource definition schemas in v1, or see the long list of [validations we wish we could tighten](https://github.com/kubernetes/kubernetes/issues/64841#issuecomment-395141013))
* Fixes to API field name typos

Those benefits aside, marking the beta resource deprecated once the GA version is available is a good signal the beta served its purpose and consumers desiring stability should use the GA API.

**Comment by thockin** [comment]
> Improvements that require changing API behavior must only be applied to
subsequent versions:

All "living" versions must be round-trippable.  So dropping from an old
version would streamline the work, but flies in the face of the rules as
written.

> Those benefits aside, marking the beta resource deprecated once the GA
version is available is a good signal the beta served its purpose and
consumers desiring stability should use the GA API.

I 100% buy this, as long as we all understand this deprecation CAN NOT lead
to piecemeal removal of the kind, unless we rewrite the deprecation rules.

In short, deprecation was written assuming that API versions would move
forward and that GVK meant (GV)K.  We've allowed that to change to G(VK),
which meshes poorly with the rules as written.

IMNSHO this is all screwed up and we should Just Fix It.  I understand it's
hard work, but holy hell, what we have evolved is a mess.

**Comment by jsafrane** [comment]
LGTM

**Comment by markyjackson-taulia** [comment]
Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this issue still intended for 1.17?

File: test/integration/etcd/data.go (hunk 2 above)
Comment (msau42) [reply test/integration/etcd/data.go]: I checked with liggitt. He clarified that there's a group default, which is set to v1, and then a per-type override, which is currently v1beta1.  For now, we keep the type override at v1beta1 for one more release, and then can remove it to use v1 in 1.18: https://github.com/kubernetes/kubernetes/issues/83513

**Comment by msau42** [comment]
All outstanding items and comments have been addressed.

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve
